Improve Transfer filter methods

bsiegel opened the pull request:
This was a pain point for many participants in the UX study. This PR presents one possible option for improving the UX.

tjprescott commented on examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift:
```diff
@@ -157,8 +157,7 @@ extension BlobDownloadViewController: UITableViewDelegate, UITableViewDataSource
         // Update background color and progress.
-        if let transfer = blobClient.transfers.downloadedFrom(container: AppConstants.videoContainer, blob: blobName)
-            .first {
+        if let transfer = blobClient.downloads.firstWith(blobName: blobName) {
```
This would be one of those situations where `firstWith` could easily lead to surprising results since, within a storage account, a blob name is unique only in combination with its container.

bsiegel replied:
It's true, I'm still not sure we're striking the right balance between user-friendliness (making it easy for the developer to find what they're looking for) and correctness (ensuring users can't introduce bugs because their assumptions are not entirely valid).

tjprescott commented on sdk/storage/AzureStorageBlob/Source/TransferManager.swift:
```diff
@@ -48,13 +48,9 @@ internal protocol TransferManager {
 
     func add(transfer: TransferImpl)
     func cancel(transfer: TransferImpl)
-    func cancelAll(withRestorationId: String?)
```
Why are these being removed?

bsiegel replied:
As part of this, a lot of the functionality we used to have in the manager is now part of either the Transfer object itself or the TransferCollection. For example, `transfer.cancel()` and `transferCollection.cancelAll()`. Thus the TransferManager protocol no longer needed to include the `*All` methods. The concrete implementation of the transfer manager still does need some implementations of them though.

bsiegel commented on sdk/storage/AzureStorageBlob/Source/DataStructures/TransferCollection.swift:
```diff
@@ -0,0 +1,156 @@
+        var matched = true
+
+        if let state = state {
+            matched = matched && transfer.state == state
```
Ah, yeah I should probably do that for each condition to make the match process fail faster.

bsiegel commented on sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift:
```diff
@@ -511,7 +512,7 @@ public final class StorageBlobClient: PipelineClient {
         toFile destinationUrl: LocalURL,
         withOptions options: DownloadBlobOptions? = nil
-    ) throws -> Transfer? {
+    ) throws -> BlobTransfer? {
```
It's not necessary for this, but it makes sense logically. `download(blob:fromContainer:toFile:)` will always start a transfer for a single blob, thus it's appropriate to return a BlobTransfer here. And that way the customer doesn't have to check the type and cast in order to see all of the available properties, all of the properties are much more discoverable this way.

Similarly, when we add a method to do a multi-blob transfer (e.g. something like `downloadAll(fromContainer:toDirectory)`) it should probably return a `MultiBlobTransfer` object.

So in general, I think that we should default to exposing the most specific type we can, since it makes it easier for the customer. And fall back to exposing a generic `Transfer` only when necessary (and the more I think about our API, the fewer places I'm finding it necessary - most places where we might mix together a collection of BlobTransfer and MultiBlobTransfer objects, we probably should instead have separate methods for getting each).

bsiegel replied:
I wanted to take another pass after this PR to actually go through and see if I could completely eliminate places where we make the user deal with the `Transfer` protocol. Internally we still use it in quite a few places, but our public API should always know whether a user-facing operation involves a single blob transfer or a multiblob transfer. 

Also, a few thoughts I had as part of the process were:

- We should really encode our states in a state machine, so that it's clear what transitions are valid and how and when they can occur. I started down this path but it was a fairly involved refactor and decided to circle back later.
- We could simplify the TransferDelegate by having only:
  - `transfer(_:didUpdateState:)`
      This is used for any state change including completion and failure. The failed state could contain an associated value with the error.
  - `transfer(_:didUpdateProgress:)`
      Called only when transfers make progress, the UX study seemed to indicate combining this with state was confusing.
- Based on the UX study feedback, should we have a delegate (or could be callbacks) that can be attached to a `Transfer` object itself, which is called only when something changes for that specific transfer? This would be in addition to, not instead of, the client delegate that can be attached to receive updates about all transfers the client creates.

tjprescott replied:
I'm okay with returning `BlobTransfer` instead of `Transfer`. The other thoughts should probably be preserved in a separate issue.

bsiegel commented on sdk/storage/AzureStorageBlob/Source/DataStructures/TransferCollection.swift:
```diff
@@ -0,0 +1,156 @@
+        blobName _: String? = nil,
+        localUrl _: LocalURL? = nil,
+        state _: TransferState? = nil
+    ) -> [BlobTransfer] {
```
Nope, that's a good catch :)

tjprescott commented on sdk/storage/AzureStorageBlob/Source/DataStructures/TransferCollection.swift:
```diff
@@ -82,11 +82,11 @@ public struct TransferCollection {
         state _: TransferState? = nil
-    ) -> [BlobTransfer] {
-        return all.filter { match($0) }
+    ) -> TransferCollection {
```
I like this. Now I can filter, get the collection, and then cancel/delete/whatever the whole result of the filter.

tjprescott commented:
Closes #311.

tjprescott commented:
@bsiegel is this ready to merge?

bsiegel commented:
@tjprescott I think this is ready to merge. I removed the `filter` and `first` methods, if people really need to do arbitrary filtering they can do so via `items.filter` instead. I also added a section to the README.
